keps/sig-api-machinery/4008-crd-ratcheting/README.md

```diff
@@ -0,0 +1,987 @@
+
+1. Users complain to CRD author. 
+2. CRD author annotates schema
+3. CRD author does not bump version
+4. Users apply updated CRD schema
+5. Controllers can successfully update the status field of the CR
+6. User's workflow keeps humming
+7. User compliments cunning CRD author for their brilliance
+8. CRD author remembers to user value validations in their schemas more in the future.
+
+#### K8s Update Tightens CRD validation
+
+In this case the user has a CRD with a field with the following schema:
+
+```yaml
+type: string
+format: dns1123subdomain
+```
+
+Kubernetes does not recognize this as a format. Today, Kubernetes treats
+[unrecognized formats as a no-op](https://github.com/kubernetes/kubernetes/blob/f9f9e7177a328ae47bac6b823d57a99dfbfd310b/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/validation/formats.go#L53-L65) and ignores them. So users would be able
+to write fields not conforming to this format to custom resources.
+
+If Kubernetes wishes to support this format, users workflows will break upon
+upgrade until the offending resources and identified and repaired.
+
+With this CRD Validation Ratcheting feature, Kubernetes would be able to
+adopt new format strings safely without breaking user workflows.
+
+#### K8s Update Widens CRD Validation
+
+In this case the user has a CRD with a field with following schema:
+
+```yaml
+type: string
+format: byte
+```
+
+In k8s 1.27 and below, this format disallows empty strings due to an implementation detail. However, in native types like `Secret`, byte is disallowed. Is Kubernetes had an update to CRDs to allow empty string for `byte`, this widens the validation for those CRDS which use it.
+
+This is fine when upgrading to new versions if Kubernetes: all existing resources still pass validation.
+
+The user may still experience issues when downgrading, for example if they had unreletated issues with the new Kubernetes update:
+
+1. User updates to new Kubernetes version
+2. New CRs that were previously disallowed are allowed to be stored
+3. User is disappointed with K8s version and downgrades
+4. Controllers fail to update some resources, since they fail validation.
+
+If the CRD Validation Ratcheting feature were enabled the following occurs:
+
+1. User updates to new Kubernetes version
+2. New CRs that were previously disallowed are allowed to be stored
+3. User is disappointed with K8s version and downgrades
+4. Controllers are allowed to update their resources, as long as all their changes are valid.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+This KEP is scoped to CRDs, but if [KEP-XXX: Declarative Validation]() is implemented then we should expect it to also affect native type validation.
+
+### Risks and Mitigations
+
+Help Wanted
```

Comment:
Maybe mention that the way a breaking schema change can be detected is different and potentially delayed since there are fewer writes that will fail?

Mitigation can be stuff like:

- use something like https://github.com/openshift/crd-schema-checker when changing CRDs to prevent accidental breaks
- rely on writes of new objects to test CRD changes.